### PR TITLE
layers: Reserve space in AS vector

### PR DESCRIPTION
**Pull request by ziga-lunarg**
This is a great performance increase, but this function can still be a bottleneck

### CODE REVIEW

File: layers/state_tracker/state_tracker.cpp [hunk 1]
```diff
@@ -6252,6 +6252,7 @@ small_vector<const vvl::AccelerationStructureKHR *, 2> DeviceState::GetAccelerat
     if (address == 0) {
         return as_vector;
     }
+    as_vector.reserve(static_cast<uint32_t>(as_with_addresses.array.size()));
```
Comment (artem-lunarg) [inline layers/state_tracker/state_tracker.cpp:6255]: What I learned recently about `small_vector` is it is indeed for small number of objects and does not work the same way as vector.

`small_vector` does not exponentially increase its size when capacity is reached. (e.g. by doubling it). It reallocates storage to have enough space for a single additional element. This results in complexity of the insert operation to be `O(n^2)` which becomes really bad for large N. Reserve is a good solution or to use a std::vector if high probability of larger number of elements.

Initially I tried to use small_vector as a performant replacement for std::vector when few elements is a common case, but gave up when found it works differently for large N.

Comment (spencer-lunarg) [reply layers/state_tracker/state_tracker.cpp:6255]: will note this comment I made Arno leave when he added this last month

```
// small_vector size comes from field experience, where because of how they recycle memory
// some games end up having the same buffer backing 2 acceleration structures,
// usually leading to them sharing the same address
small_vector<const vvl::AccelerationStructureKHR*, 2> GetAccelerationStructuresByAddress(VkDeviceAddress address) const;
```

File: layers/state_tracker/state_tracker.cpp
Comment (spencer-lunarg) [inline layers/state_tracker/state_tracker.cpp:6263]: sorry, I see this now, ya doing the reserve make sense... my question is maybe for @arno-lunarg 

Once we return the array here, do we ever adjust it?

**Comment by ziga-lunarg** [comment]
> > This is a great performance increase, but this function can still be a bottleneck
> 
> So I guess what were you testing? CTS?
> 
> I have nothing to do but believe you, but I would like @arno-lunarg (who I am pretty sure wrote this) to weigh in first

Yes a CTS test was stuck here forever, the reserve() improved the time, but it is still slow. However I believe what the CTS test is doing is an edge case that wont be hit often

**Comment by spencer-lunarg** [comment]
>  However I believe what the CTS test is doing is an edge case that wont be hit often

overall, looking at it, the `reserve` makes sense IMO
